2022-05-24 Future-based shuffle for Ray.md

```diff
@@ -0,0 +1,108 @@
+- For non-shuffle operators, this interface returns `[chunk.key]`
+- For the shuffle operator, this interface returns `[(chunk.key, reducer_idx1), ...... , (chunk.key, reducer_idxn)]
+  - For shuffle operators such as DataFrameGroupByOperand, the number of output shuffle blocks is fixed, and each mapper outputs `N` blocks, so all data keys`[(chunk.key, reducer_idx1), ...... , (chunk.key, reducer_idxn)]`
+  - For shuffle operators such as `TensorReshape`, each mapper may not have the outputs to all reducers, and the shuffle data keys it outputs is determined at runtime. For this type of operator, we need to return all data keys corresponding to all reducers`[(chunk.key, reducer_idx1), ...... , (chunk.key, reducer_idxn)]`; then when reducer iter_mapper_data, skip some partitions if there is no data for key. 
+  ![image](https://user-images.githubusercontent.com/12445254/169472129-f78828d7-5855-4386-837c-19208f848dcc.png)
+
+
+This solution has some drawbacks:
+- The metadata overhead would be very large, which will lead to supervisor OOM and task lineage evict. Assuming that we want to support `100T` data shuffle, the data size of each input chunk is 1GB, and the number of mappers and reducers is `100,000` respectively, then the data keys of each subtask will occupy about 4M memory, as shown in Figure 1, plus `FetchShuffle` stores source_keys/source_idxes/source_mappers, the data will expand by 3 times, that is, a single subtask will occupy `16M` memory. If the supervisor memory limit is 16G, this will cause the ray backend OOM when submitting 1000 subtasks.
+![image](https://user-images.githubusercontent.com/12445254/169472798-535b8229-8d82-473b-82a4-26136270a9c7.png)
+- Introduce shuffle operands implement burden because `get_data_keys` needs to be impelmented. 
+
+### Building shuffle DAG by reducer index
+Acutally, Mars' ChunkGraph has information about the upstream and downstream chunk dependencies. After converting to Ray DAG, this part of the information still exists. The reason for relying on data keys is that Mars replaced `ShuffleProxy` with `FetchShuffle`. The implementation of `Fetch/FetchShuffle` is an abstraction of P2P, which is actually inconsistent with the semantics of Ray DAG. Therefore, when using Ray Backend, a new set of Shuffle operators can be introduced to remove FetchShuffle and rely on Ray to automatically resolve subtask input arguments.
+
+#### Operand Changes
+- Add `n_reducers` and `reducer_ordinal` to all `MapReduceOperand` operands
+- Introduce a new group of `FetchShuffle` operands which only records `n_mappers` and `n_reducers` to reduce meta overhead:
+```python
+class FetchShuffleType(Enum):
+    FETCH_SHUFFLE_BY_INDEX = 0
+    FETCH_SHUFFLE_BY_DATA_KEY = 1
+
+class FetchShuffleByIndex(Operand):
+    _op_type_ = opcodes.FETCH_SHUFFLE_BY_INDEX
+    n_mappers = Int32Field("n_mappers")
+    n_reducers = Int32Field("n_reducers")
```

Comment:
Better to clarify how `n_mappers` and `n_reducers` locates input mappers and output reducers. BTW, `n_xxx` refers to number of certain objects, thus it is better to give a new name.

Reply:
I think `n_mappers` and `n_reducers` are OK. There are similar namign such as `n_workers`/`n_threads`

Reply:
What's the key of this operand and related chunk? Some clarification needed.

Reply:
It's similar to `FetchShuffle` operands, the only difference is that it didn't include `source_keys/source_idxes/source_mappers` info and added `n_mappers `/`n_reducers `

Reply:
For me, it is more convenient to include a key of `ShuffleProxy` operand to reference the whole shuffle group.

Reply:
You mean set the key of `ShuffleProxy` to `FetchShuffleByIndex` operand? Sounds acceptable to me.